include/triton/core/tritonbackend.h

```diff
@@ -1255,6 +1255,67 @@ TRITONBACKEND_ModelInstanceReportBatchStatistics(
+/// successfully. In this case all timestamps should be non-zero values reported
+/// in nanoseconds and should be collected using
+/// std::chrono::steady_clock::now().time_since_epoch() or the equivalent.  Set
+/// 'success' to false to indicate that the inference response failed to
+/// complete successfully. In this case all timestamps values are ignored.
+///
+/// For consistency of measurement across different backends, the
+/// timestamps should be collected at the following points during
+/// TRITONBACKEND_ModelInstanceExecute.
+///
+///   TRITONBACKEND_ModelInstanceExecute()
+///     CAPTURE TIMESPACE (response_start)
+///     < when inference response object is created >
+///     CAPTURE TIMESPACE (compute_output_start)
+///     < allocate output buffers and extract output tensors, including
+///       copyting the tensors to/from GPU if necessary >
+///     CAPTURE TIMESPACE (response_end)
+///     return
+///
+/// Note that these statistics are associated with a valid
+/// TRITONBACKEND_Response object and so must be reported before the response is
+/// sent. This API can be used both with models that return a single response or
+/// multiple responses for each request. If the model performs a single
+/// inference and then provides multiple resposnes, it is recommended to account
+/// for the inference time in the request statistics instead of response
+/// statistics. However, if there is an inference time associated with every
+/// response, it must be accounted for in the response statistics.
+///
+/// \param instance The model instance.
+/// \param response The inference response that statistics are being
+/// reported for.
+/// \param success True if the inference response completed
+/// successfully, false if it failed to complete.
+/// \param response_start Timestamp for the start of execution
+/// computations.
+/// \param compute_output_start Timestamp for the end of execution
+/// computations.
+/// \param send_flags The flags that will be used for sending the response.
+/// \param response_end Timestamp for the end of execution.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceReportResponseStatistics(
+    TRITONBACKEND_ModelInstance* instance, TRITONBACKEND_Response* response,
+    const bool success, const uint64_t response_start,
+    const uint64_t compute_output_start, const uint64_t response_end,
+    const uint32_t response_flags);
+
+/// Record statistics for an inference with no response.
+///
+/// This API is used with models that return no response. If a model returns
+/// one or more responses, the
+/// TRITONBACKEND_ModelInstanceReportResponseStatistics should be used instead.
+///
+/// \param instance The model instance.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceReportNoResponseStatistics(
```

Comment:
@tanmayv25 / @kthui What if the response is using only flags but it is not the final response? It looks like the API doesn't permit reporting statistics for this scenario? This interface should be changed to:

```
TRITONBACKEND_ModelInstanceReportResponseFactoryStatistics(
    TRITONBACKEND_ModelInstance* instance, TRITONBACKEND_ResponseFactory* response_factory,
    const bool success, const uint64_t response_start,
    const uint64_t compute_output_start, const uint64_t response_end,
    const uint32_t response_flags);
```

Reply:
@Tabrizian, please help me understand. Do you mean a scenario which a model can dynamically choose how many response(s) to return?

Given a model that will return at most 3 responses, per request.
```
| res 1 | -> | res 2 | -> | res 3 |
```
In the normal case, all three responses are computed and returned.
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ✅          ✅
```
For some requests, the model does not return the second response.
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ❓          ✅
```
Are you saying we should report statistics (i.e. compute time) for the second response? (where there is not a response object)

-----
If this is correct, maybe we can simply add a new metric to the response statistics, from `success`, `fail`, `compute_infer` and `compute_output` to `success`, `fail`, *`no_output`*, `compute_infer` and `compute_output`?

For example, if the model returns the following three responses:
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ❓          ❌
```
Then, `res 1` is `success`, `res 2` is `no_output` and `res 3` is `fail`, and `compute_infer` and `compute_output` are calculated.

The newly added `response_stats` can be simplified into, with the above example: 
```
"response_stats": [
        {
                "index": 0,
                "success": {
                        "count": 1,
                        "ns": 10
                },
                "fail": {
                        "count": 0,
                        "ns": 0
                },
                "no_output": {
                        "count": 0,
                        "ns": 0
                },
                "compute_infer": {
                        "count": 1,
                        "ns": 4
                },
                "compute_output": {
                        "count": 1,
                        "ns": 5
                }
        }, 
        {
                "index": 1,
                "success": {
                        "count": 0,
                        "ns": 0
                },
                "fail": {
                        "count": 0,
                        "ns": 0
                },
                "no_output": {
                        "count": 1,
                        "ns": 8
                },
                "compute_infer": {
                        "count": 1,
                        "ns": 4
                },
                "compute_output": {
                        "count": 0,
                        "ns": 0
                }
        },
        {
                "index": 2,
                "success": {
                        "count": 0,
                        "ns": 10
                },
                "fail": {
                        "count": 1,
                        "ns": 10
                },
                "no_output": {
                        "count": 0,
                        "ns": 0
                },
                "compute_infer": {
                        "count": 1,
                        "ns": 4
                },
                "compute_output": {
                        "count": 1,
                        "ns": 5
                }
        }
]
```
By sending more requests, the respective `count` will increase correspondingly and the `ns` will update.
-----
CC @tanmayv25

Reply:
> Are you saying we should report statistics (i.e. compute time) for the second response? (where there is not a response object)

Right, I'm saying that it looks like there is no way to report statistics for the second response in your example. I think the current backend API is too specific and should be extended to cover this scenario.

> If this is correct, maybe we can simply add a new metric to the response statistics, from success, fail, compute_infer and compute_output to success, fail, no_output, compute_infer and compute_output?

How about changing the HTTP/GRPC statistics output to something like below:

```
"response_stats": [
        {
                "index": 2,
                 response: {
                  "success": {
                          "count": 0,
                          "ns": 10
                  },
                  "fail": {
                          "count": 1,
                          "ns": 10
                  },
              
                  "compute_infer": {
                          "count": 1,
                          "ns": 4
                  },
                  "compute_output": {
                          "count": 1,
                          "ns": 5
                  },
                  "flag": "TRITONSERVER_RESPONSE_COMPLETE_FINAL"
                 },
                 response: {
                  "success": {
                          "count": 0,
                          "ns": 10
                  },
                  "fail": {
                          "count": 1,
                          "ns": 10
                  },
              
                  "compute_infer": {
                          "count": 1,
                          "ns": 4
                  },
                  "compute_output": {
                          "count": 1,
                          "ns": 5
                  },
                  "flag": "0"
                 }
        }
]
```

This would allow the API to be more extensible in case we add new flags to the response object.

Reply:
Can you elaborate your through on the flag object a bit more? For example, what are some possible flags to be added in the future. It is because currently the flag is only used to detect if the provided response is the last one for this request.

There might be another scenario to be considered:
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ❓          ✅
```
and
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ✅          ❓
```
Both requests generate two responses, but the API cannot distinguish between them. Then, we might want the user to provide a statistics for `res 2` and `res 3` regardless whether a response is returned. Do you mean using the flag object to distinguish it?
If so, do we still need an array of `response_stats`, like `[2 responses grouped here], [3 responses grouped here]`? It is because there will always be 3 responses in the above scenario.

Reply:
I see. Looks like flag doesn't resolve the issue. How about renaming "flag" to "is_flag_only" which indicates whether a response is only flags and this will be set to true for responses reported using `TRITONBACKEND_ModelInstanceReportResponseFactoryStatistics`? For other response statistics this will be set to false?

Reply:
I was thinking maybe we can simplify the response statistics to a table.

For example, the following responses are generated:
```
request 1: 
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ❓          ✅

request 2:
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ✅          ✅

request 3:
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ✅          ❓

request 4:
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ✅          ❌
```

We can aggregate by the `res #` for each request.

For example, before any request, the response statistics:
|  | Success | Fail | No output | Compute infer | Compute output |
| --- | --- | --- | --- | --- | --- |

After request 1:
|  | Success | Fail | No output | Compute infer | Compute output |
| --- | --- | --- | --- | --- | --- |
| res 1 | 1 | 0 | 0 | 1 | 1 |
| res 2 | 0 | 0 | 1 | 1 | 0 |
| res 3 | 1 | 0 | 0 | 1 | 1 |

After request 2:
|  | Success | Fail | No output | Compute infer | Compute output |
| --- | --- | --- | --- | --- | --- |
| res 1 | 2 | 0 | 0 | 2 | 2 |
| res 2 | 1 | 0 | 1 | 2 | 1 |
| res 3 | 2 | 0 | 0 | 2 | 2 |

After request 3:
|  | Success | Fail | No output | Compute infer | Compute output |
| --- | --- | --- | --- | --- | --- |
| res 1 | 3 | 0 | 0 | 3 | 3 |
| res 2 | 2 | 0 | 1 | 3 | 2 |
| res 3 | 2 | 0 | 1 | 3 | 2 |

After request 4:
|  | Success | Fail | No output | Compute infer | Compute output |
| --- | --- | --- | --- | --- | --- |
| res 1 | 4 | 0 | 0 | 4 | 4 |
| res 2 | 3 | 0 | 1 | 4 | 3 |
| res 3 | 2 | 1 | 1 | 3 | 2 |

Note: only count is shown above, ns average should be included as well.

The backend implementation must report statistics for each possible response under the appropriate category. The core will aggregate the reports into response statistics, available for query.

I think this should make the response statistics much simpler to understand, but I do not know if I missed any reportable metrics that is covered by the current design.

Reply:
I like the idea being proposed by @kthui. It hides the complexity of flags from the reported statistics. The onus is on backend developer to properly send the compute times for each response. The intent of statistics report should be how much work the backend expended on these request/response processing. 
I think we can remove `TRITONBACKEND_ModelInstanceReportNoResponseStatistics`. And ask users to properly set response_start, compute_output_starts and response_end timestamps in the `TRITONBACKEND_ModelInstanceReportResponseStatistics` itself for empty responses?

Reply:
> I think we can remove `TRITONBACKEND_ModelInstanceReportNoResponseStatistics`. And ask users to properly set response_start, compute_output_starts and response_end timestamps in the `TRITONBACKEND_ModelInstanceReportResponseStatistics` itself for empty responses?

Yes. For grouping, we only need the `res #` for each call to `TRITONBACKEND_ModelInstanceReportResponseStatistics`.

I think we can either figure it out from the `request` objects' unique id and with the order of calling, or simply request the backend to provide the `res #` with each call.

Reply:
@Tabrizian what do you think about the change away from flag?

Reply:
> I think this should make the response statistics much simpler to understand, but I do not know if I missed any reportable metrics that is covered by the current design.

The very first proposal was similar to what you describe here but the problem is that it is not possible find the answer to the questions below:

1. What is the average request time for requests that generate 4 responses?
2. How long does the third response take in requests that generate 4 responses?

These questions can be answered in the current design.

> I think we can remove TRITONBACKEND_ModelInstanceReportNoResponseStatistics. And ask users to properly set response_start, compute_output_starts and response_end timestamps in the TRITONBACKEND_ModelInstanceReportResponseStatistics itself for empty responses?

The idea of `TRITONBACKEND_ModelInstanceReportResponseFactoryStatistics` is similar to what you described here. It is exactly like the other API except it accepts a response factory object instead of a response object.

> For grouping, we only need the res # for each call to TRITONBACKEND_ModelInstanceReportResponseStatistics.

I think we should try to hide the response number if we can. It would make the API easier to use.

Reply:
I see. An outer dimension to the table is needed to track statistics for requests that generate different number of responses. Then, it is very similar to the design you proposed.

Reply:
Jumping into the discussion without knowing the whole picture here:
```
| res 1 | -> | res 2 | -> | res 3 |
   ✅           ❓          ✅
```
What does this case even mean here? Why we want to add index on to the responses? Shouldn't this collapse to the case that the model generates 2 responses for the request? From the client's perspective, it only receives 2 responses at the end.

Note that "flag" is not a response, it is utility for Triton to manipulate the sequence of responses (now the only use case is where the response sequence end). Response is the way to convey messages, and if user does want to communicate data other than tensors, they should be able to describe it as one of the Triton model output. In this sense, res 2 will still be in the form of Triton response and then the example will be case the the model generates 3 responses for the request. On the client side they will read and identify that res 2 is some special response.

Reply:
> What does this case even mean here? Why we want to add index on to the responses? Shouldn't this collapse to the case that the model generates 2 responses for the request? From the client's perspective, it only receives 2 responses at the end.

IMO, the way that response to client relationship is currently handled is just one implementation of the Triton Server API. I also don't know exactly what this means or if there are use-cases for it but the API allows this. If customers are using this right now, they can't report statistics for this scenario.

 > Why we want to add index on to the responses?
 
 It is a way to distinguish the responses (e.g.. the first response took this time).